src/wrappers.jl

```diff
@@ -0,0 +1,65 @@
+  :(Base.ReinterpretArray{T,N,<:Src})             => (A,mut)->Base.reinterpret(eltype(A), mut(parent(A))),
+  :(LinearAlgebra.Adjoint{T,<:Dst})               => (A,mut)->LinearAlgebra.adjoint(mut(parent(A))),
+  :(LinearAlgebra.Transpose{T,<:Dst})             => (A,mut)->LinearAlgebra.transpose(mut(parent(A))),
+  :(LinearAlgebra.LowerTriangular{T,<:Dst})       => (A,mut)->LinearAlgebra.LowerTriangular(mut(parent(A))),
+  :(LinearAlgebra.UnitLowerTriangular{T,<:Dst})   => (A,mut)->LinearAlgebra.UnitLowerTriangular(mut(parent(A))),
+  :(LinearAlgebra.UpperTriangular{T,<:Dst})       => (A,mut)->LinearAlgebra.UpperTriangular(mut(parent(A))),
+  :(LinearAlgebra.UnitUpperTriangular{T,<:Dst})   => (A,mut)->LinearAlgebra.UnitUpperTriangular(mut(parent(A))),
+  :(LinearAlgebra.Diagonal{T,<:Dst})              => (A,mut)->LinearAlgebra.Diagonal(mut(parent(A))),
+  :(LinearAlgebra.Tridiagonal{T,<:Dst})           => (A,mut)->LinearAlgebra.Tridiagonal(mut(A.dl), mut(A.d), mut(A.du)),
+)
+
+for (W, ctor) in _wrappers
+    mut = :(A -> adapt(to, A))
+    @eval adapt_structure(to, wrapper::$W where {T,N,Src,Dst}) = $ctor(wrapper, $mut)
+end
+
+"""
+    WrappedArray{T,N,Src,Dst}
+
+Union-type that encodes all array wrappers known by Adapt.jl. Typevars `T` and `N` encode
+the type and dimensionality of the resulting container.
+
+Two additional typevars are used to encode the parent array type: `Src` when the wrapper
+uses the parent array as a source, but changes its properties (e.g.
+`SubArray{T,1,Array{T,2}` changes `N`), and `Dst` when those properties are copied and thus
+are identical to the destination wrapper's properties (e.g. `Transpose{T,Array{T,N}}` has
+the same dimensionality as the inner array). When creating an alias for this type, e.g.
+`WrappedSomeArray{T,N} = WrappedArray{T,N,...}` the `Dst` typevar should typically be set to
+`SomeArray{T,N}` while `Src` should be more lenient, e.g., `SomeArray`.
+
+Only use this type for dispatch purposes. To convert instances of an array wrapper, use
+[`adapt`](@ref).
+"""
+const WrappedArray{T,N,Src,Dst} = @eval Union{$([W for (W,ctor) in Adapt._wrappers]...)} where {T,N,Src,Dst}
+
+# XXX: this Union is a hack:
+# - only works with one level of wrappi ng
+# - duplication of Src and Dst typevars (without it, we get `WrappedGPUArray{T,N,AT{T,N}}`
+#   not matching `SubArray{T,1,AT{T,2}}`, and leaving out `{T,N}` makes it impossible to
+#   match e.g. `Diagonal{T,AT}` and get `N` out of that). alternatively, computed types
+#   would make it possible to do `SubArray{T,N,<:AT.name.wrapper}` or `Diagonal{T,AT{T,N}}`.
+#
+# ideally, Base would have, e.g., `Transpose <: WrappedArray`, and we could use
+# `Union{SomeArray, WrappedArray{<:Any, <:SomeArray}}` for dispatch.
+# https://github.com/JuliaLang/julia/pull/31563
+
+# accessors for extracting information about the wrapper type
+Base.ndims(::Type{<:WrappedArray{T,N,Src,Dst}}) where {T,N,Src,Dst} = @isdefined(N) ? N : ndims(Dst)
```

Comment:
This type-piracy is unfortunate. @jakebolewski and I just hunted this down as the source of breaking GPU inference, because Julia doesn't inline this function (might be the unused Src type parameter, e.g. dangling type parameter).

why is the `@defined` needed here?

x-ref: https://github.com/CliMA/ClimateMachine.jl/pull/1260#issuecomment-646788859

Reply:
```
julia> f(a) = eltype(a)
f (generic function with 1 method)

julia> @code_typed f(B)
CodeInfo(
1 ─     return Float64
) => Type{Float64}

julia> using Adapt
[ Info: Precompiling Adapt [79e6a3ab-5dfb-504d-930d-738a2a938a0e]

julia> @code_typed f(B)
CodeInfo(
1 ─ %1 = invoke Main.eltype(_2::SubArray{Float64,1,Array{Float64,2},Tuple{Base.Slice{Base.OneTo{Int64}},Int64},true})::Core.Compiler.Const(Float64, false)
└──      return %1
) => Type{Float64}
```

Reply:
> why is the `@defined` needed here?

```
julia> x(::Base.ReshapedArray{Float32,2,Array{Float32,2}})
ERROR: syntax: invalid "::" syntax around REPL[3]:1
Stacktrace:
 [1] top-level scope at REPL[3]:1

julia> x(Base.ReshapedArray{Float32,2,Array{Float32,2}})
(Float32, 2)

julia> x(LinearAlgebra.Adjoint{Float32,Array{Float32,2}})
ERROR: UndefVarError: N not defined
Stacktrace:
 [1] x(::Type{Adjoint{Float32,Array{Float32,2}}}) at ./REPL[2]:1
 [2] top-level scope at REPL[8]:1
```

For types that 'fix' the N typevar I can't get a hold of it from the WrappedArray union...
It's all a bit of a kludge, admittedly, but it brings us closer to trying out something that might end up in Base some day.

Suggested fix? `Adapt.eltype` and `Adapt.ndims`?

Reply:
I had not seen https://github.com/JuliaGPU/Adapt.jl/pull/26

Reply:
I think making these `Adapt.eltype` would make sense